Firestore/core/src/firebase/firestore/remote/serializer.h

```diff
@@ -111,7 +116,56 @@ class Serializer {
   firebase::firestore::model::DocumentKey DecodeKey(
       absl::string_view name) const;
 
+  /**
+   * @brief Converts the Document (i.e. key/value) into bytes.
+   *
+   * @param[out] out_bytes A buffer to place the output. The bytes will be
+   * appended to this vector.
+   * @return A Status, which if not ok(), indicates what went wrong. Note that
+   * errors during encoding generally indicate a serious/fatal error.
+   */
+  // TODO(rsgowman): Similar to above, if we never support any output except to
+  // a vector, it may make sense to have Serializer own the vector and provide
+  // an accessor rather than asking the user to create it first.
+  util::Status EncodeDocument(
+      const firebase::firestore::model::DocumentKey& key,
+      const firebase::firestore::model::ObjectValue& value,
+      std::vector<uint8_t>* out_bytes) const;
+
+  /**
+   * @brief Converts from bytes to the model Document format.
+   *
+   * @param bytes The bytes to convert. These bytes must represent a
+   * BatchGetDocumentsResponse. It's assumed that exactly all of the bytes will
+   * be used by this conversion.
+   * @return The model equivalent of the bytes or a Status indicating
+   * what went wrong.
+   */
+  util::StatusOr<std::unique_ptr<model::MaybeDocument>> DecodeMaybeDocument(
```

Comment:
Another (slight) asymmetry. This time, we're returning a MaybeDocument, but we're taking in bytes that represent a BatchGetDocumentsResponse (rather than bytes that represent a MaybeDocument.) However, there *is* no MaybeDocument proto object; the closest analogue is BatchGetDocumentsResponse, so this seems reasonable. (And of course, there's never any guarantee that the network things will have exact analogues with the model things.)

Re naming: Something that occurs to me with the nanopb port: We've been naming the methods Decode*Type*. But is 'Type' the parameter type or return type or proto type or model type? Generally, all 4 are roughly the same, so it usually doesn't matter. But here, they differ. The other ports use modeltype, which In the other ports was fine, since the proto type showed up in the parameter list (or return value). But here, BatchGetDocumentsResponse is represented in the parameter list as `uint8_t* bytes`; the only way you'd know what message it is is to read the docs. (And who has time for that?)

Possible solution is to rename everything to `[En|De]code*ProtoType*`